### DISCUSSION TITLE
Task<T> covariance

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#5841] [Language Ideas] [answered]
I'm pretty sure this was discussed somewhere but I couldn't find anything, there is https://github.com/dotnet/csharplang/issues/3950 but that only talks about NRTs.
```cs
Task<List<int>> Foo() { ... }
Task<IEnumerable<int>> Bar()
{
  return Foo(); // error
}
```
I'm pretty sure the problem is that `Task<T>` is a class, not an interface, and only interfaces support variance (and arrays/delegates). Is it really that difficult to make this work?
Could I make my own `ITask<T>` interface, to be task-like, which is implemented by a wrapper (`TaskWrapper<T>`) around existing `Task<T>`?

PS. I know I can just make these `async` and use `await` and it works, but that introduces state machine, etc. - trying to avoid this, if possible.

PPS. Similar issues arise with `FromResult`:
```cs
protected override Task<IEnumerable<int>> Foo()
{
  return Task.FromResult(Array.Empty<int>()); // error
  return Task.FromResult(Array.Empty<int>().CastTo<IEnumerable<int>>()); // have to do this currently
}
```

**Comment by pinkfloydx33** [comment, marked as answer]
https://github.com/dotnet/csharplang/discussions/2498

**Comment by CyrusNajmabadi** [comment]
You should be able to do this:

```c#
return Task.FromResult<IEnumerable<int>>(Array.Empty<int>());
```

**Reply by TahirAhmadov** [reply]
I know, still, would be nice not to type all that.
